haskell.compiler.*: assert that host->target == build->target tools

sternenseemann opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

`targetCC` is used to populate the environment variables telling GHC's
configure script where to find the tools the resulting GHC will use to
compile and link compilation units. For this the tools passed via
misleadingly named (i. e. like what would usually be the build/native
tools) end up in GHC's settings file to be looked up at runtime.

Consequently, these tools need to run on the host platform and not the
build platform as before with targetCC coming from toolsForTarget. Since
it is currently impossible to cross-compile GHC in nixpkgs (i. e. the
build and host platform are always the same), this change doesn't make a
difference in practice for the moment (and also hasn't caused any build
failures in the past).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Ericson2314 commented:
@sternenseemann Err now we are getting into the issue that GHC's outer build system requires build == host. Those commands are run at build time too, and don't just end up in the config file.

Ericson2314 commented:
(Oops, meant just to leave comment, not close this!)

sternenseemann commented:
> @sternenseemann Err now we are getting into the issue that GHC's outer build system requires build == host. Those commands are run at build time too, and don't just end up in the config file.

Right, I was a bit tired yesterday and for some reason assumed that `config.mk.in` would only relate to `settings`… It seems that the intention is to rely on `PATH` for both the build->target and later the host->target tools (which are assumed to be named the same).

So if we want to cross-compile GHC at some point, we'd need to mess with `settings` after installing as things stand, swapping out the left over build->target tools from compiling with the proper host->target tools. For this the `ghc-settings` tool idea of @angerman would be useful indeed.

For now, I'd then propose that we just add an `assert pkgsHostTarget.targetPackages.stdenv.cc == pkgsBuildTarget.targetPackages.stdenv.cc`?

sternenseemann committed:
haskell.compiler.*: use targetCC for hasGold check

This is a bit shorter and more consistent with the rest of the file.

sternenseemann committed:
haskell.compiler.*: assert that host->target == build->target tools

CC, CXX, LD, AR, …, LLC, OPT and CLANG will be invoked by GHC's build
system at build time in the build->target role. However, since we are
passing absolute paths, they will get saved in GHC's settings file and
later invoked at runtime, when they should be host->target. This means
that the build->target and host->target tools need to be the same for
our built GHC to work properly which is what we guard using these new
asserts.

Being able to drop these asserts would be a step towards cross-compiling
GHC (as opposed to building a GHC cross-compiler which still works).

sternenseemann commented:
@Ericson2314 okay, changed this to add a few asserts which make the current assumptions clearer (NB: these assumptions didn't get introduced in #147260, but were already there before since e. g. the build->target `CC`, `LD` etc. would leak into settings and thus into the resulting GHC's run time).
